File: bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/dataset/DatasetRecord.java
package edu.gemini.spModel.dataset;

/**
 * Combination of DatasetQaRecord and DatasetExecRecord.
 */
public final class DatasetRecord {
    public final DatasetQaRecord qa;
    public final DatasetExecRecord exec;

    public DatasetRecord(DatasetQaRecord qa, DatasetExecRecord exec) {
        if (qa == null) throw new NullPointerException();
        if (exec == null) throw new NullPointerException();
        if (!qa.label.equals(exec.getLabel())) {
            final String msg = String.format("QA and Exec records for different datasets: %s vs %s", qa.label, exec.getLabel());
            throw new IllegalArgumentException(msg);
        }
        this.qa   = qa;
        this.exec = exec;
    }

    public DatasetLabel getLabel() { return qa.label; }

    public DatasetRecord withQa(DatasetQaRecord qa) { return new DatasetRecord(qa, exec); }
    public DatasetRecord withExec(DatasetExecRecord exec) { return new DatasetRecord(qa, exec); }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        final DatasetRecord that = (DatasetRecord) o;

        if (!exec.equals(that.exec)) return false;
        if (!qa.equals(that.qa)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = qa.hashCode();
        result = 31 * result + exec.hashCode();
        return result;
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder("DatasetRecord{");
        sb.append("qa=").append(qa);
        sb.append(", exec=").append(exec);
        sb.append('}');
        return sb.toString();
    }
}

File: bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/dataset/DatasetRecord.scala
package edu.gemini.spModel.dataset

import scalaz._
import Scalaz._

/** Combination `DatasetQaRecord` and `DatasetExecRecord`.  These are the
  * user-editable and server-updated (respectively) bits of a dataset.
  */
final case class DatasetRecord(qa: DatasetQaRecord, exec: DatasetExecRecord) {

  def label: DatasetLabel = qa.label

  // This is code converted from Java in order to get an extractor that matches
  // the name of the class.  This is poor but these tests were being applied
  // in the Java code and need to be reworked into the model to prevent these
  // situations.

  import Implicits._
  assert(qa   != null, "Cannot construct a DatasetRecord with a null DatasetQaRecord")
  assert(exec != null, "Cannot construct a DatasetRecord with a null DatasetExecRecord")
  assert(qa.label === exec.label, s"QA and Exec records for different datasets: ${qa.label} vs ${exec.label}")

  // Used by Java code.
  def withQa(qa: DatasetQaRecord): DatasetRecord       = copy(qa = qa)
  def withExec(exec: DatasetExecRecord): DatasetRecord = copy(exec = exec)
}

object DatasetRecord {
  implicit val EqualDatasetRecord: Equal[DatasetRecord] = Equal.equalA
}
